src/ripple/shamap/impl/SHAMapInnerNode.cpp

```diff
@@ -33,26 +33,718 @@
+        static_cast<void (*)(void*)>(boost::singleton_pool<
+                                     boost::fast_pool_allocator_tag,
+                                     arrayChunkSizeBytes[I],
+                                     boost::default_user_allocator_new_delete,
+                                     std::mutex,
+                                     chunksPerBlock[I],
+                                     chunksPerBlock[I]>::free)...,
+    };
+}
+std::array<std::function<void(void*)>, boundaries.size()> const freeArrayFuns =
+    initFreeArrayFuns(std::make_index_sequence<boundaries.size()>{});
+
+template <std::size_t... I>
+std::array<std::function<bool(void*)>, boundaries.size()> initIsFromArrayFuns(
+    std::index_sequence<I...>)
+{
+    return std::array<std::function<bool(void*)>, boundaries.size()>{
+        boost::singleton_pool<
+            boost::fast_pool_allocator_tag,
+            arrayChunkSizeBytes[I],
+            boost::default_user_allocator_new_delete,
+            std::mutex,
+            chunksPerBlock[I],
+            chunksPerBlock[I]>::is_from...,
+    };
+}
+std::array<std::function<bool(void*)>, boundaries.size()> const
+    isFromArrayFuns =
+        initIsFromArrayFuns(std::make_index_sequence<boundaries.size()>{});
+
+// This function returns an untagged pointer
+[[nodiscard]] inline std::pair<std::uint8_t, void*>
+allocateArrays(std::uint8_t numChildren)
+{
+    auto const i = boundariesIndex(numChildren);
+    return {i, allocateArrayFuns[i]()};
+}
+
+// This function takes an untagged pointer
+inline void
+deallocateArrays(std::uint8_t boundaryIndex, void* p)
+{
+    assert(isFromArrayFuns[boundaryIndex](p));
+    freeArrayFuns[boundaryIndex](p);
+}
+#else
+
+template <std::size_t... I>
+std::array<std::pmr::synchronized_pool_resource, boundaries.size()>
+    initPmrArrayFuns(std::index_sequence<I...>)
+{
+    return std::array<std::pmr::synchronized_pool_resource, boundaries.size()>{
+        std::pmr::synchronized_pool_resource{std::pmr::pool_options{
+            .max_blocks_per_chunk = chunksPerBlock[I],
+            .largest_required_pool_block = chunksPerBlock[I]}}...,
+    };
+}
+std::array<std::pmr::synchronized_pool_resource, boundaries.size()>
+    pmrArrayFuns =
+        initPmrArrayFuns(std::make_index_sequence<boundaries.size()>{});
+
+// This function returns an untagged pointer
+[[nodiscard]] inline std::pair<std::uint8_t, void*>
+allocateArrays(std::uint8_t numChildren)
+{
+    auto const i = boundariesIndex(numChildren);
+    return {i, pmrArrayFuns[i].allocate(arrayChunkSizeBytes[i])};
+}
+
+// This function takes an untagged pointer
+inline void
+deallocateArrays(std::uint8_t boundaryIndex, void* p)
+{
+    return pmrArrayFuns[boundaryIndex].deallocate(
+        p, arrayChunkSizeBytes[boundaryIndex]);
+}
+#endif
+
+[[nodiscard]] inline int
+popcnt16(std::uint16_t a)
+{
+#if defined(__clang__) || defined(__GNUC__)
+    return __builtin_popcount(a);
```

Comment:
@nbougalis Looks like the `__builtin_popcount` is faster than my "fallback" table lookup even when I don't use `-march`. Now I'm interested in what algorithm they used. Interesting that using `-march` actually isn't much faster. This is likely because the constant benchmark overhead is such a significant portion of the timing. Given this, it probably isn't worth jumping through hoops to enable `popcnt` for this case (tho we should still look at what speedups we can get in other places).

Also, just to defend myself, I _did_ know that that it wouldn't use the `popcnt` assembler instruction unless we used `-march` :-)


Popcount with `-march`
```
-------------------------------------------------------------------
Benchmark                         Time             CPU   Iterations
-------------------------------------------------------------------
BM_popcnt_builtin              6.98 ns         6.98 ns    100474566
BM_popcnt_lookup               12.3 ns         12.3 ns     57152228
```
Popcount without `-march`
```
------------------------------------------------------------
Benchmark                  Time             CPU   Iterations
------------------------------------------------------------
BM_popcnt_builtin       7.80 ns         7.80 ns     88919230
BM_popcnt_lookup        12.2 ns         12.2 ns     57612945
```
FYI: Here's a gist of the benchmark I used: https://gist.github.com/seelabs/b22db88a9b6c539a9fe1980cffbcd7e6

Reply:
I peeked at what gcc is using as a fallback if `popcnt` isn't available. It's a series of "logical shift right" followed by "add with carry". Since C++ doesn't have access to the "add with carry" instruction, we can't use this algorithm without using intrinsics (or assembler).

I also searched for other `popcount` implementations. Here's one that's almost as fast as the gcc fallback:
```
[[nodiscard]] inline int
popcount(std::uint32_t v)
{
    v = v - ((v >> 1) & 0x55555555); 
    v = (v & 0x33333333) + ((v >> 2) & 0x33333333);
    return ((v + (v >> 4) & 0xF0F0F0F) * 0x1010101) >> 24;
}
```
When I benchmark that I get:
```
---------------------------------------------------------------
Benchmark                     Time             CPU   Iterations
---------------------------------------------------------------
BM_popcnt_builtin          7.81 ns         7.80 ns     90264669
BM_popcnt_lookup           12.1 ns         12.1 ns     57758234
BM_popcnt_bittwiddle       8.00 ns         8.00 ns     87404423
```

However, that's a little too magic for my tastes. I think the table lookup is a good enough fallback.

Reply:
Yeah, [Bit Twiddling Hacks](https://graphics.stanford.edu/~seander/bithacks.html) are awesome, but I don't know that I'd like to see them in the `rippled` codebase in particular. Again, awesome, but it's the sort of esoteric wizard code whose performance is highly-dependent on the CPU it runs on and it's written so that that nobody understands it, except a guy who used to program the PDP/11 that was in the basement and is now on this 3rd retirement.

Also, the performance of the `__builtin_popcount` fall-back implementation is only marginally slower than hardware operation. It would take a 1.25 billion invocations to "shave" a second off and that's if all we ever did was count bits.